geom/dcel_extract.go

```diff
@@ -0,0 +1,295 @@
+				// during linear and point geometry extraction.
+				edge.label |= extracted
+				edge.twin.label |= extracted
+				edge.origin.label |= extracted
+
+				if seen[edge] {
+					return
+				}
+				if include(edge.twin.incident.label) {
+					// Adjacent face is in the polygon, so this edge cannot be part
+					// of the boundary.
+					seen[edge] = true
+					return
+				}
+				seq := extractPolygonBoundary(facesInPoly, edge, seen)
+				ring, err := NewLineString(seq)
+				if err != nil {
+					panic(fmt.Sprintf("could not create LineString: %v", err))
+				}
+				rings = append(rings, ring)
+			})
+		}
+
+		// Construct the polygon.
+		orderCCWRingFirst(rings)
+		poly, err := NewPolygonFromRings(rings)
+		if err != nil {
+			return nil, err
+		}
+		polys = append(polys, poly)
+	}
+	return polys, nil
+}
+
+func extractPolygonBoundary(faceSet map[*faceRecord]bool, start *halfEdgeRecord, seen map[*halfEdgeRecord]bool) Sequence {
+	var coords []float64
+	e := start
+	for {
+		seen[e] = true
+		xy := e.origin.coords
+		coords = append(coords, xy.X, xy.Y)
+
+		// Sweep through the edges around the vertex (in a counter-clockwise
+		// order) until we find the next edge that is part of the polygon
+		// boundary.
+		e = e.twin.prev.twin
+		for !faceSet[e.incident] {
+			e = e.prev.twin
+		}
+
+		if e == start {
+			break
+		}
+	}
+	coords = append(coords, coords[:2]...)
+	return NewSequence(coords, DimXY)
+}
+
+// findFacesMakingPolygon finds all faces that belong to the polygon that
+// contains the start face (according to the given inclusion criteria).
+func findFacesMakingPolygon(include func(uint8) bool, start *faceRecord) map[*faceRecord]bool {
+	expanded := make(map[*faceRecord]bool)
+	toExpand := make(map[*faceRecord]bool)
+	toExpand[start] = true
+
+	for len(toExpand) > 0 {
+		var popped *faceRecord
+		for f := range toExpand {
+			delete(toExpand, f)
+			popped = f
+			break
+		}
+
+		adj := adjacentFaces(popped)
+		expanded[popped] = true
+		for _, f := range adj {
+			if !include(f.label) {
+				continue
+			}
+			if expanded[f] {
+				continue
+			}
+			if toExpand[f] {
+				continue
+			}
+			toExpand[f] = true
+		}
+	}
+	return expanded
+}
+
+// orderCCWRingFirst reorders rings such that if it contains at least one CCW
+// ring, then a CCW ring is the first element.
+func orderCCWRingFirst(rings []LineString) {
+	for i, r := range rings {
+		if ccw := signedAreaOfLinearRing(r, nil) > 0; ccw {
+			rings[i], rings[0] = rings[0], rings[i]
+			return
+		}
+	}
+}
+
+// TODO: Line extraction isn't working too well at the moment. It's currently
+// extracting each line individually, which isn't intended. It might be better
+// to return a []line here, and then construct back into LineString and
+// MultiLineString as a separate logical step since it seems tricky to do
+// inline.
+
+func (d *doublyConnectedEdgeList) extractLineStrings(include func(uint8) bool) ([]LineString, error) {
```

Comment:
Given that the extraction algorithm here doesn't work as intended, we can replace it with something simpler.

Reply:
Fixed.